configV2/etcd.go

```
@@ -87,42 +101,156 @@ func NewETCDConfig(vdiskID string, endpoints []string) (ConfigSource, error) {
+					if len(ev.Kv.Value) < 1 {
+						cfg.nbdRWLock.Lock()
+						cfg.nbd = nil
+						cfg.nbdRWLock.Unlock()
+						continue
+					}
+
+					// create subconfig and check validity
+					cfg.baseRWLock.RLock()
+					newNBD, err := newNBDConfig(ev.Kv.Value, cfg.vdiskID, cfg.base.Type)
+					if err != nil {
+						log.Errorf("Did not update %s of %s: %s", cfg.nbdKey, cfg.vdiskID, err)
+						cfg.baseRWLock.RUnlock()
+						continue
+					}
+					cfg.baseRWLock.RUnlock()
+					cfg.nbdRWLock.Lock()
+					cfg.nbd = newNBD
+					cfg.nbdRWLock.Unlock()
+				case cfg.tlogKey:
+					if len(ev.Kv.Value) < 1 {
+						cfg.tlogRWLock.Lock()
+						cfg.tlog = nil
+						cfg.tlogRWLock.Unlock()
+						continue
+					}
+
+					newTlog, err := newTlogConfig(ev.Kv.Value)
+					if err != nil {
+						log.Errorf("Did not update %s of %s: %s", cfg.tlogKey, cfg.vdiskID, err)
+						continue
+					}
+					cfg.tlogRWLock.Lock()
+					cfg.tlog = newTlog
+					cfg.tlogRWLock.Unlock()
+				case cfg.slaveKey:
+					if len(ev.Kv.Value) < 1 {
+						cfg.slaveRWLock.Lock()
+						cfg.slave = nil
+						cfg.slaveRWLock.Unlock()
+						continue
+					}
+
+					newSlave, err := newSlaveConfig(ev.Kv.Value)
+					if err != nil {
+						log.Errorf("Did not update %s of %s: %s", cfg.slaveKey, cfg.vdiskID, err)
+						continue
+					}
+					cfg.slaveRWLock.Lock()
+					cfg.slave = newSlave
+					cfg.slaveRWLock.Unlock()
+				}
+			}
+		}
+	}()
+}
+
 // Close implements ConfigSource.Close
 func (cfg *etcdConfig) Close() error {
 	// close client
 	err := cfg.cli.Close()
 	if err != nil {
 		return err
 	}
+
 	return nil
 }
 
 // Base implements ConfigSource.Base
 func (cfg *etcdConfig) Base() BaseConfig {
+	cfg.baseRWLock.RLock()
+	defer cfg.baseRWLock.RUnlock()
 	return cfg.base
 }
 
 // NBD implements ConfigSource.NBD
 func (cfg *etcdConfig) NBD() (*NBDConfig, error) {
+	cfg.nbdRWLock.RLock()
+	defer cfg.nbdRWLock.RUnlock()
+	if cfg.nbd == nil {
+		return nil, nil
```

This should return an error. Whoever is calling `NBD` is clearly assuming there is an NBD config available. In this case, no NBD config available, and thus that is an error. If you are going to start returning `nil` even when `error` is `nil`, you always need to do 2 checks, instead of 1. There are cases where this is desired, but this is not one of those. You can make a "constant" error `ErrConfigNotAvailable` for this one (or whatever the name convention is).